### ISSUE TITLE
IDISP003: False positive when disposable used as a ref parameter.

### ISSUE THREAD

**Issue by AraHaan** [DotNetAnalyzers/IDisposableAnalyzers#267]
I noticed an false positive (stated from issue #202):

```cs
namespace AnythingWorksHere
{
    using System;
    using System.ComponentModel;
    using System.Diagnostics;
    using System.Diagnostics.CodeAnalysis;
    using System.Drawing;
    using System.Drawing.Drawing2D;
    using System.Runtime.InteropServices;
    using System.Windows.Forms;
    using Els_kom.Enums;
    using Elskom.Generic.Libs;

    internal class ThemedForm : Form
    {
        private bool minimizeHover;
        private bool maximizeHover;
        private bool closeHover;
        private bool helpHover;
        private bool minimizeClicked;
        private bool maximizeClicked;
        private bool closeClicked;
        private bool helpClicked;
        private bool active;
        private bool mouseDown;
        [SuppressMessage("Usage", "CA2213:Disposable fields should be disposed", Justification = "Gets disposed of automatically.")]
        private ContextMenuStrip systemMenuStrip = null;
        private Point lastLocation;
        private bool menuShown;
        private int oldIcon;
        private Timer iconSyncTmr;

        // state bitmaps.
        private Bitmap minimizeBitmap;
        private Bitmap minimizeHoverBitmap;
        private Bitmap minimizeDisabledBitmap;
        private Bitmap maximizeBitmap;
        private Bitmap maximizeHoverBitmap;
        private Bitmap maximizeDisabledBitmap;
        private Bitmap closeBitmap;
        private Bitmap closeHoverBitmap;
        private Bitmap closeClickedBitmap;
        private Bitmap helpBitmap;
        private Bitmap helpHoverBitmap;
        private Bitmap iconBitmap;
        private Icon lastIcon;

        internal ThemedForm()
        {
            this.InitializeComponent();
            this.Components = new Container();

            // These seems to be the correct locations for the caption buttons in Windows 10.
            this.MinimizeHitbox = new Rectangle(this.Size.Width - 136, 1, 45, 29);
            this.MaximizeHitbox = new Rectangle(this.Size.Width - 90, 1, 45, 29);
            this.CloseHitbox = new Rectangle(this.Size.Width - 44, 1, 45, 29);
            this.HelpHitbox = new Rectangle(this.Size.Width - 90, 1, 45, 29);

            // these are fixed points never changeing.
            this.IconHitbox = new Rectangle(9, 7, 16, 16);

            // the caption in the title bar. We need this to set the bit to make form moveable again.
            this.CaptionHitbox = new Rectangle(0, 0, this.Size.Width, 31);
            // this.Paint += this.ThemedForm_Paint;
            this.Load += this.ThemedForm_Load;
            // this.MouseUp += this.ThemedForm_MouseUp;
            // this.MouseDown += this.ThemedForm_MouseDown;
            // this.MouseMove += this.ThemedForm_MouseMove;
            // this.MouseLeave += this.ThemedForm_MouseLeave;
            // this.Activated += this.ThemedForm_Activated;
            // this.Deactivate += this.ThemedForm_Deactivate;
            if (ShareXResources.Theme != ShareXTheme.GetPresets()[0])
            {
                ShareXResources.Theme = ShareXTheme.GetPresets()[0];
            }

            // create state bitmaps.
            this.CreateBitmap(ref this.minimizeBitmap, this.MinimizeHitbox, true, false);
            this.CreateBitmap(ref this.minimizeHoverBitmap, this.MinimizeHitbox, true, true);
            this.CreateBitmap(ref this.minimizeDisabledBitmap, this.MinimizeHitbox, false, false);
            this.CreateBitmap(ref this.maximizeBitmap, this.MaximizeHitbox, true, false);
            this.CreateBitmap(ref this.maximizeHoverBitmap, this.MaximizeHitbox, true, true);
            this.CreateBitmap(ref this.maximizeDisabledBitmap, this.MaximizeHitbox, false, false);
            this.CreateBitmap(ref this.closeBitmap, this.CloseHitbox, true, false);
            this.CreateBitmap(ref this.closeHoverBitmap, this.CloseHitbox, true, true);
            this.CreateBitmap(ref this.closeClickedBitmap, this.CloseHitbox, true, false, false, true);
            this.CreateBitmap(ref this.helpBitmap, this.HelpHitbox, true, false, false);
            this.CreateBitmap(ref this.helpHoverBitmap, this.HelpHitbox, true, true, false);
            this.CreateBitmap(ref this.iconBitmap, this.IconHitbox, true, false);
        }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        [EditorBrowsable(EditorBrowsableState.Advanced)]
        [Localizable(false)]
        public Container Components { get; } = null;

        // Unrelated Windows Forms stuff.

        protected override Size DefaultSize => new(300, 300);

        private Rectangle MinimizeHitbox { get; set; }

        private Rectangle MaximizeHitbox { get; set; }

        private Rectangle CloseHitbox { get; set; }

        private Rectangle HelpHitbox { get; set; }

        private Rectangle CaptionHitbox { get; set; }

        private Rectangle IconHitbox { get; set; }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        /// <param name="disposing">true if managed resources should be disposed; otherwise, false.</param>
        [SuppressMessage("IDisposableAnalyzers.Correctness", "IDISP007:Don't dispose injected.", Justification = "Needed to not leak the memory from the Icons.")]
        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                this.Components?.Dispose();
                this.minimizeBitmap?.Dispose();
                this.minimizeHoverBitmap?.Dispose();
                this.minimizeDisabledBitmap?.Dispose();
                this.maximizeBitmap?.Dispose();
                this.maximizeHoverBitmap?.Dispose();
                this.maximizeDisabledBitmap?.Dispose();
                this.closeBitmap?.Dispose();
                this.closeHoverBitmap?.Dispose();
                this.closeClickedBitmap?.Dispose();
                this.helpBitmap?.Dispose();
                this.helpHoverBitmap?.Dispose();
                this.iconBitmap?.Dispose();
                this.iconSyncTmr?.Dispose();

                // this.lastIcon?.Dispose()
            }

            base.Dispose(disposing);
        }

        // Unrelated Windows Forms stuff.

        private void ThemedForm_Load(object sender, EventArgs e)
        {
            this.FormBorderStyle = FormBorderStyle.None;
            foreach (Control control in this.Controls)
            {
                var tmp = new Point(control.Location.X, control.Location.Y);
                tmp.X += 1;
                tmp.Y += 31;
                control.Location = tmp;
            }

            if (!this.DesignMode)
            {
                // Yes I used some code from ShareX to provide theming to my forms.
                // ShareXResources.ApplyTheme(this);
                this.iconSyncTmr.Enabled = true;
            }
        }

        // Unrelated Windows forms stuff.

        // lazily create the bitmap.
        [SuppressMessage("IDisposableAnalyzers.Correctness", "IDISP007:Don't dispose injected.", Justification = "Needed to cleanup old bitmap.")]
        private void CreateBitmap(ref Bitmap bitmap, Rectangle bounds, bool enabled, bool hovered, bool isMaximize = true, bool isClicked = false)
        {
            if (bitmap == this.iconBitmap && Icons.IconEquals(this.lastIcon, this.Icon))
            {
                return;
            }

            bitmap?.Dispose();
            bitmap = new Bitmap(bounds.Width, bounds.Height);
            var bmpRect = new Rectangle(0, 0, bounds.Width, bounds.Height);

            // in case we got to draw the buttons to the bitmap!!!
            // the 2 colors for the close button was taken from the default windows 10 theme file.
            using var pen = new Pen(ShareXResources.Theme.TextColor);
            using var brush = new LinearGradientBrush(bmpRect, ShareXResources.Theme.MenuHighlightColor, ShareXResources.Theme.MenuHighlightColor, LinearGradientMode.Vertical);
            using var closeHoverBrush = new LinearGradientBrush(bmpRect, Color.FromArgb(255, 232, 17, 35), Color.FromArgb(255, 232, 17, 35), LinearGradientMode.Vertical);
            using var closeClickedBrush = new LinearGradientBrush(bmpRect, Color.FromArgb(153, 231, 16, 34), Color.FromArgb(153, 231, 16, 34), LinearGradientMode.Vertical);
            using var gr = Graphics.FromImage(bitmap);
            if (bounds == this.IconHitbox)
            {
                this.oldIcon = SettingsFile.SettingsJson.WindowIcon;
                this.lastIcon = this.Icon;
                using var iconbmp = this.Icon.ToBitmap();
                using var bmp = new Bitmap(iconbmp, 16, 16);
                gr.DrawImage(bmp, new Point(0, 0));
            }
            else if (bounds == this.MinimizeHitbox)
            {
                if (enabled)
                {
                    if (hovered)
                    {
                        gr.FillRectangle(brush, bmpRect);
                    }

                    gr.DrawLine(pen, new Point(18, 15), new Point(28, 15));
                }
                else
                {
                    gr.DrawLine(Pens.DarkGray, new Point(18, 15), new Point(28, 15));
                }
            }
            else if ((bounds == this.MaximizeHitbox) && isMaximize)
            {
                if (enabled)
                {
                    if (hovered)
                    {
                        gr.FillRectangle(brush, bmpRect);
                    }

                    gr.DrawRectangle(pen, 18, 10, 10, 10);
                }
                else
                {
                    // color it with Pens.LightGray.
                    gr.DrawRectangle(Pens.DarkGray, 18, 10, 10, 10);
                }
            }
            else if (bounds == this.CloseHitbox)
            {
                if (hovered)
                {
                    gr.FillRectangle(closeHoverBrush, bmpRect);
                }
                else if (isClicked)
                {
                    gr.FillRectangle(closeClickedBrush, bmpRect);
                }

                // TODO: Draw icon image of the button itself like windows does.
                gr.DrawImage(Properties.Resources.closeglyph, new Point(18, 10));
            }
            else if (bounds == this.HelpHitbox)
            {
                if (hovered)
                {
                    gr.FillRectangle(brush, bmpRect);
                }

                // TODO: Draw icon image of the button itself like windows does.
                gr.DrawImage(Properties.Resources.helpglyph, new Point(18, 10));
            }
        }

        private void IconSyncTmr(object sender, EventArgs e)
        {
            var iconVal = SettingsFile.SettingsJson.WindowIcon;
            if (!iconVal.Equals(this.oldIcon))
            {
                // reset the icon displayed here.
                this.CreateBitmap(ref this.iconBitmap, this.IconHitbox, true, false);
                this.Invalidate(this.IconHitbox);

                // finally change the value.
                this.oldIcon = iconVal;
            }
        }

        private void InitializeComponent()
        {
            this.SuspendLayout();

            // iconSyncTmr
            this.iconSyncTmr?.Dispose();
            this.iconSyncTmr = new Timer(this.Components)
            {
                Enabled = false,
                Interval = 1,
            };
            this.iconSyncTmr.Tick += new EventHandler(this.IconSyncTmr);

            // ThemedForm
            this.ClientSize = new Size(300, 300);
            this.FormBorderStyle = FormBorderStyle.None;
            this.Name = "ThemedForm";
            this.ResumeLayout(false);
        }
    }
}
```

Some of the code references projects from https://github.com/Elskom/Sdk/, for that I would recommend building the packages and then have it use the MetaPackage (Elskom.Sdk which pulls in the analyzer automatically as well).
Note: I plan to eventually setup package uploading to pr builds, and to nuget eventually on what I have changed from before, also to clone it it would have to be done recursively.

**Comment by JohanLarsson** [comment]
I'm curious, why ref and not return?

```cs
private void CreateBitmap(ref Bitmap bitmap) { ... }
```

vs

```cs
private Bitmap CreateBitmap() { ... return bitmap; }
```

Analyzer can do a better job with the ref case.

**Comment by JohanLarsson** [comment]
Ok, I see what is going on here. This will be tricky to analyze for the general case, we would need to find all call sites and check that it is safe to dispose the ref argument. For public methods this will be impossible.
I think I'll close this as wontfix as writing the method so that it returns a created instance is trivially analyzed and easier for human readers.

It is one of the goals for this analyzer to nudge us into writing simpler code. My thinking is that if the analyzer struggles we should pause and think if we can refactor the code into something simpler. How do you feel about his?

**Comment by AraHaan** [comment]
alright.

**Comment by JohanLarsson** [comment]
Thank you for the issue!
